{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open(\"covid19_to_other_6_sars_consensus.csv\",\"r+\") as f:\n",
    "    data = [x.replace(\" \",\"\").replace(\"[]\",\"\").replace(\"\\n\",\"\") for x in f.readlines()]\n",
    "f.close()\n",
    "\n",
    "data= [x.split(\",\") for x in data]\n",
    "data = data[0]\n",
    "data = data[1:]\n",
    "\n",
    "data=[float(x) for x in data]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in positions of reference sequence (or else won't align correctly!)\n",
    "with open(\"NC_045512_alignment_sequence.txt\",\"r+\") as f:\n",
    "    ref_seq =[x.replace(\"\\n\",\"\") for x in f.readlines()]\n",
    "f.close()\n",
    "ref_seq = (\"\".join(ref_seq)) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remap sequence positions\n",
    "# remove positions associated with \"-\" regions\n",
    "data_new = []\n",
    "# ref_seq = list(ref_seq)\n",
    "posn_nums = []\n",
    "i=0\n",
    "j=0\n",
    "while i<len(ref_seq):\n",
    "    if(ref_seq[i]==\"-\"):\n",
    "        data_new.append(0)\n",
    "        posn_nums.append(-1)\n",
    "    else:\n",
    "        data_new.append(data[i])\n",
    "        (posn_nums.append(j+1))\n",
    "        j+=1\n",
    "    i+=1\n",
    "\n",
    "data=data_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31976\n",
      "31976\n",
      "31976\n",
      "29903\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# below two values should be the same\n",
    "print(len(ref_seq))\n",
    "print(len(data))\n",
    "print(len(posn_nums))\n",
    "\n",
    "\n",
    "print(len(ref_seq.replace(\"-\",\"\"))) # should be 29903 for NC_045512.2\n",
    "# check that sequences match\n",
    "with open(\"NC_045512.fasta\",\"r+\") as f:\n",
    "    seq =f.readlines()[0][:-1]\n",
    "f.close()\n",
    "print(seq==ref_seq.replace(\"-\",\"\")) # should be equal\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getIndexPositions(listOfElements, element):\n",
    "    ''' Returns the indexes of all occurrences of give element in\n",
    "    the list- listOfElements '''\n",
    "    indexPosList = []\n",
    "    indexPos = 0\n",
    "    while True:\n",
    "        try:\n",
    "            # Search for item in list from indexPos to the end of list\n",
    "            indexPos = listOfElements.index(element, indexPos)\n",
    "            # Add the index position in list\n",
    "            indexPosList.append(indexPos)\n",
    "            indexPos += 1\n",
    "        except ValueError as e:\n",
    "            break\n",
    " \n",
    "    return indexPosList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_ranges(start,stop,data,gene,cutoff_pcntl,posn_nums):\n",
    "    req_region_num = 12\n",
    "    \n",
    "    seq_len = 16\n",
    "    \n",
    "    start = posn_nums.index(start)\n",
    "    stop = posn_nums.index(stop)\n",
    "\n",
    "    # create sliding window of seq_len and identify regions where average % identity is high\n",
    "    import statistics \n",
    "    avg_idenity_ls = []\n",
    "    index_ls = []\n",
    "    # homology % cutoff\n",
    "    hm = 80\n",
    "    ct = 6\n",
    "    while(len(index_ls) < 3):\n",
    "        index_ls = []\n",
    "        i = start\n",
    "        while i<stop:\n",
    "            # positions 1-7 homology > hm %\n",
    "            count = len([i for i in (data[i:i+7]) if i >= hm]) \n",
    "            if(count>ct):\n",
    "                # positions 1-16 homology > hm-20 % (less stringent for rest of sequence)\n",
    "                count = len([i for i in (data[i+7:i+seq_len]) if i >= hm-20])\n",
    "                if(count>ct):\n",
    "#                 print(posn_nums[i])\n",
    "#                 print(data[i:i+7])\n",
    "                    index_ls.append(posn_nums[i])#+start])\n",
    "            i+=1\n",
    "        hm-=5\n",
    "        ct-=0.5\n",
    "        \n",
    "        \n",
    "        \n",
    "    \n",
    "\n",
    "    \n",
    "    output_data_str = (\"gene:\"+gene+\" | homology cutoff :\"+str(hm+5)+\" | number of regions  \"+\n",
    "                       str(len(index_ls))+\n",
    "                       \" | total number of regions: \"+str(i))\n",
    "    \n",
    "    \n",
    "    # TODO: determine actual positions (not just in alignment)\n",
    "    \n",
    "    \n",
    "    return(output_data_str,index_ls)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# protein,start position,end position\n",
    "gene_data =[['pp1a',266,13483],\n",
    "            ['pp1ab',266,21555,],\n",
    "            [\"spike\",21563,25384],\n",
    "            [\"3a\",25393,26220],\n",
    "            [\"envelope\",26245,26472],\n",
    "            [\"matrix\",26523,27191],\n",
    "            [\"7a\",27394,27759],\n",
    "            [\"8b\",27894,28259],\n",
    "            [\"nucleocapsid\",28274,29533]\n",
    "           ]\n",
    "\n",
    "output_data = []\n",
    "for g in gene_data:\n",
    "    output_data.append(compute_ranges(g[1],g[2],data,g[0],99,posn_nums))\n",
    "\n",
    "# output to file\n",
    "with open(\"range_outputs.txt\",\"w+\") as f:\n",
    "    i = 0 \n",
    "    while i<len(output_data): # loop through each gene\n",
    "        j = 0\n",
    "        f.write(output_data[i][0])\n",
    "        f.write(\"\\n\")\n",
    "        while j<len(output_data[i][1]):\n",
    "            f.write(str(output_data[i][1][j]).replace(\"[\",\"\").replace(\"]\",\"\"))\n",
    "            f.write(\"\\n\")\n",
    "            j+=1\n",
    "        i+=1\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# def compute_ranges(start,stop,data,gene,cutoff_pcntl,posn_nums):\n",
    "#     req_region_num = 12\n",
    "    \n",
    "#     seq_len = 16\n",
    "    \n",
    "#     start = posn_nums.index(start)\n",
    "#     stop = posn_nums.index(stop)\n",
    "\n",
    "#     print(start)\n",
    "#     print(stop)\n",
    "#     data = data[start-1:stop]\n",
    "#     indexes = list(range(1,len(data)))\n",
    "#     data = [data,indexes]\n",
    "#     # create sliding window of seq_len and identify regions where average % identity is high\n",
    "#     import statistics \n",
    "#     avg_idenity_ls = []\n",
    "#     index_ls = []\n",
    "#     i = 0\n",
    "#     while i<(len(data[0])-seq_len):\n",
    "# #         avg_idenity_ls.append(statistics.mean(data[0][i:i+seq_len]))\n",
    "#         avg_idenity_ls.append(sum(data[0][i:i+seq_len]))\n",
    "\n",
    "#         index_ls.append(i+1)\n",
    "#         i+=1\n",
    "\n",
    "#     plt.boxplot(avg_idenity_ls)\n",
    "\n",
    "\n",
    "#     import numpy as np\n",
    "#     num_regions_abv_co=0\n",
    "    \n",
    "#     too_few_regions_ct = 0\n",
    "    \n",
    "#     while(num_regions_abv_co < req_region_num):\n",
    "#         percentile = np.percentile(avg_idenity_ls,cutoff_pcntl)\n",
    "#         # count number of regions above percentile\n",
    "#         top_regions = ([x > percentile for x in avg_idenity_ls])\n",
    "        \n",
    "#         # if not enough regions (req_region_num) then repeat decreasing cutoff_pcntl by 1 each time\n",
    "#         num_regions_abv_co =len([x for x in avg_idenity_ls if x > percentile])\n",
    "#         print([x for x in avg_idenity_ls if x > percentile])\n",
    "#         too_few_regions_ct+=1\n",
    "#         cutoff_pcntl-=1\n",
    "        \n",
    "# #     if(too_few_regions_ct<2):\n",
    "# #         print(\"HERE\")\n",
    "# #         while((num_regions_abv_co > req_region_num) and cutoff_pcntl<100) :\n",
    "# # #             print(cutoff_pcntl)\n",
    "# #             percentile = np.percentile(avg_idenity_ls,cutoff_pcntl)\n",
    "# #             # count number of regions above percentile\n",
    "# #             top_regions = ([x > percentile for x in avg_idenity_ls])\n",
    "# #             # if too many regions (req_region_num) then repeat increase cutoff_pcntl by 1 each time\n",
    "# #             num_regions_abv_co =len([x for x in avg_idenity_ls if x > percentile])\n",
    "# #             print(percentile)\n",
    "# #             cutoff_pcntl+=0.1\n",
    "    \n",
    "# #     # increase % identity\n",
    "# #     too_few_regions_ct = 0\n",
    "# #     percentile = np.percentile(avg_idenity_ls,cutoff_pcntl)\n",
    "# #     while(num_regions_abv_co < req_region_num):\n",
    "# #         # count number of regions above percentile\n",
    "# #         top_regions = ([x > percentile for x in avg_idenity_ls])\n",
    "# #         # if not enough regions (req_region_num) then repeat decreasing cutoff_pcntl by 1 each time\n",
    "# #         num_regions_abv_co =len([x for x in avg_idenity_ls if x > percentile])\n",
    "# #         too_few_regions_ct+=1\n",
    "# #         percentile-=1\n",
    "        \n",
    "# # #     if(too_few_regions_ct<2):\n",
    "# # #         while((num_regions_abv_co > req_region_num+20) and cutoff_pcntl<100) :\n",
    "# # #             # count number of regions above percentile\n",
    "# # #             top_regions = ([x > percentile for x in avg_idenity_ls])\n",
    "# # #             # if too many regions (req_region_num) then repeat increase cutoff_pcntl by 1 each time\n",
    "# # #             num_regions_abv_co =len([x for x in avg_idenity_ls if x > percentile])\n",
    "# # #             percentile+=1\n",
    "    \n",
    "#     output_data_str = (\"gene:\"+gene+\" | number of regions above \"+\n",
    "#                        str(cutoff_pcntl+1)+\"th percentile (\"+\n",
    "#                        str(percentile)+\"% identity): \"+\n",
    "#                        str(num_regions_abv_co)+\n",
    "#                        \" | total number of regions: \"+str(len(avg_idenity_ls)))\n",
    "    \n",
    "#     high_homol_posns = getIndexPositions(top_regions,True)\n",
    "#     # add in distance start to account for position of each gene\n",
    "#     high_homol_posns = [x+start for x in high_homol_posns]\n",
    "#     # get regions with high homology\n",
    "#     # if nearest neighbor is also in group just jump to next\n",
    "#     regions = []\n",
    "#     prev_pos = high_homol_posns[0]\n",
    "#     i = 0\n",
    "#     while i<len(high_homol_posns):\n",
    "#         p=high_homol_posns[i]\n",
    "#         if(len(regions) == 0): # adds first element\n",
    "#             regions.append(p)\n",
    "#         else:\n",
    "#             if(prev_pos+1 != p): # if not consecutive position add this and the most previous position\n",
    "#                 regions.append(\"-\")\n",
    "#                 regions.append(high_homol_posns[i-1]) # add most previous position\n",
    "#                 regions.append(p) # add current position\n",
    "#         prev_pos = p\n",
    "#         i+=1\n",
    "#     i =0\n",
    "#     new_regions = []\n",
    "#     while i<len(regions):\n",
    "#         if(i+1 >= len(regions)):\n",
    "#             new_regions.append([regions[i]])\n",
    "#             i+=1\n",
    "#         elif(regions[i+1] == \"-\"):\n",
    "#             new_regions.append([regions[i],regions[i+2]+seq_len]) # adds seq_len bases to include whole region\n",
    "#             i+=3\n",
    "#         else:\n",
    "#             new_regions.append([regions[i]])\n",
    "#             i+=1\n",
    "\n",
    "#     # add extra seq_len to ones with single element to get region\n",
    "#     new_regions_2 = []\n",
    "#     i=0\n",
    "#     while i<len(new_regions):\n",
    "#         if len(new_regions[i]) == 1:\n",
    "#             new_regions_2.append([new_regions[i][0],new_regions[i][0]+seq_len])\n",
    "#         else:\n",
    "#             new_regions_2.append(new_regions[i])\n",
    "#         i+=1\n",
    "            \n",
    "    \n",
    "\n",
    "#     return(output_data_str,new_regions_2)  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
